Implemet most of the database functions

kalbasit opened the pull request:
# implements
- notmuch_database_get_path
- notmuch_database_needs_upgrade
- notmuch_database_upgrade
- notmuch_database_begin_atomic and notmuch_database_end_atomic as one function with a callback to call
- notmuch_database_add_message
- notmuch_database_remove_message
- notmuch_database_find_message
- notmuch_database_find_message_by_filename
- notmuch_database_get_all_tags
- notmuch_database_compact

# does not implement
- notmuch_database_get_revision

zenhack commented on db.go:
```diff
@@ -57,44 +60,153 @@ func Open(path string, mode DBMode) (*DB, error) {
+	if err := statusErr(cerr); err != nil {
+		return err
+	}
+	return nil
```
Everything after the callback could be collapsed to:

```
return statusErr(C.notmuch_database_end_atomic(db.cptr))
```

zenhack commented on db.go:
```diff
@@ -57,44 +60,153 @@ func Open(path string, mode DBMode) (*DB, error) {
 
-func (db *DB) toC() *C.notmuch_database_t {
-	return (*C.notmuch_database_t)(db.cptr)
+// Path return the database path of the database.
```
s/return/returns/

zenhack commented on db.go:
```diff
@@ -57,44 +60,153 @@ func Open(path string, mode DBMode) (*DB, error) {
+
+// AddMessage adds a new message to the current database or associate an
+// additional filename with an existing message.
+func (db *DB) AddMessage(p string) (*Message, error) {
```
Maybe change the parameter name to `filename`. It's more descriptive, and matches notmuch to boot.

zenhack commented on db.go:
```diff
@@ -57,44 +60,153 @@ func Open(path string, mode DBMode) (*DB, error) {
+	defer C.free(unsafe.Pointer(cpath))
+
+	msg := &Message{}
+	cmsg := (**C.notmuch_message_t)(unsafe.Pointer(msg.cptr))
```
This looks incorrect to me; I think you need to be taking the address of `msg.cptr`.

zenhack commented on db.go:
```diff
@@ -57,44 +60,153 @@ func Open(path string, mode DBMode) (*DB, error) {
+	return msg, nil
+}
+
+// FindMessageByFilename finds a message with the given message_id.
```
s/mesage_id/filename/

zenhack commented on errors.go:
```diff
@@ -0,0 +1,81 @@
+
+var (
+	// ErrOutOfMemory is returned when an Out of memory occured.
+	ErrOutOfMemory = statusErr(C.NOTMUCH_STATUS_OUT_OF_MEMORY)
```
Not sure we should expose this one; we should just be catching these and panicing.

kalbasit replied:
I kept this one exposed and `checkOOM` now panic with it so we can compare it to the error received in recover.

zenhack commented:
Re: removing checkOOM, the comments in `notmuch.h` state pretty clearly that this function will return NULL on OOM; we need to catch that. Not sure why you think it can be removed?

Other than the comments I made this looks pretty good to me. Good call on removing `toC()`; it was an artifact left over from a previous design which I'd discarded.

kalbasit commented:
@zenhack addressed all of your comments. In regards to `checkOOM` I brought it back but I made it panic `ErrOutOfMemory` so the implementer has a chance to programmatically find out the cause of the panic (as opposed to a random string)

kalbasit commented:
@zenhack also I have no clue how to satisfy a `char**` argument in C. I'd like to implement `notmuch_database_get_revision` but I wasn't able to yesterday. Can you please write it up for me?

Signature for `notmuch_database_get_revision`:

```
/**
 * Return the committed database revision and UUID.
 *
 * The database revision number increases monotonically with each
 * commit to the database.  Hence, all messages and message changes
 * committed to the database (that is, visible to readers) have a last
 * modification revision <= the committed database revision.  Any
 * messages committed in the future will be assigned a modification
 * revision > the committed database revision.
 *
 * The UUID is a NUL-terminated opaque string that uniquely identifies
 * this database.  Two revision numbers are only comparable if they
 * have the same database UUID.
 */
unsigned long
notmuch_database_get_revision (notmuch_database_t *notmuch,
                const char **uuid);
```

zenhack commented:
>    @zenhack also I have no clue how to satisfy a char*\* argument in C.
>    I'd like to implement notmuch_database_get_revision but I wasn't able
>    to yesterday. Can you please write it up for me?

Will do. I'll take a second look at the rest as soon as I can.

kalbasit commented:
cool. I just implemented `notmuch_database_compact` and we have only revision left.

zenhack commented on db_test.go:
```diff
@@ -66,3 +69,160 @@ func TestLastStatus(t *testing.T) {
+	}
+	defer db.Close()
+
+	// same test as TestFindMessage to ensure the database integrity.
```
Probably should factor the two out into a common function then.

kalbasit replied:
I usually don't do that in testing as I'd like the test to fully tell the story but I'm open to debate :)

zenhack commented:
I hit a local test failure because NOTMUCH_STATUS_PATH_ERROR doesn't exist in the latest release; it's only on master (which is what travis has been testing). I changed the `.travis.yml` to test against the release branch, could you rebase off of master and fix this?

Other than that it looks good.

I'll note that there is more than just revision left; most of the tags stuff isn't there yet (but of course that needn't be part of this pr).

kalbasit commented:
@zenhack IRC?

kalbasit commented:
@zenhack tags stuff implemented in #7
